File: apps/companion/src/migrate.ts
import { basename, join } from "node:path";
import { type Database, sql as database } from "./db";

const migrationsDirectory = join(import.meta.dir, "..", "migrations");

export interface MigrationFile {
  version: string;
  sql: string;
}

export async function readMigrationFiles(): Promise<MigrationFile[]> {
  const glob = new Bun.Glob("*.sql");
  const names = Array.from(
    glob.scanSync({ cwd: migrationsDirectory, onlyFiles: true }),
  ).sort();

  return Promise.all(
    names.map(async (name) => ({
      version: basename(name, ".sql"),
      sql: await Bun.file(join(migrationsDirectory, name)).text(),
    })),
  );
}

export async function runMigrations(sql: Database): Promise<void> {
  await sql.unsafe(
    "CREATE TABLE IF NOT EXISTS schema_migrations (version text PRIMARY KEY, applied_at timestamptz NOT NULL DEFAULT now())",
  );

  const appliedRows = await sql<{ version: string }[]>`
    SELECT version
    FROM schema_migrations
  `;
  const applied = new Set(appliedRows.map((row) => row.version));

  for (const migration of await readMigrationFiles()) {
    if (applied.has(migration.version)) {
      continue;
    }

    await sql.begin(async (transaction) => {
      await transaction.unsafe(migration.sql);
      await transaction`
        INSERT INTO schema_migrations (version)
        VALUES (${migration.version})
      `;
    });
  }
}

if (import.meta.main) {
  try {
    await runMigrations(database);
  } finally {
    await database.end();
  }
}

File: apps/companion/src/migrate.rs
use std::collections::HashSet;

use sqlx::PgPool;

const MIGRATIONS: &[(&str, &str)] = &[
    (
        "0001_foundation",
        include_str!("../migrations/0001_foundation.sql"),
    ),
    ("0002_chunks", include_str!("../migrations/0002_chunks.sql")),
    (
        "0003_observation_dedupe",
        include_str!("../migrations/0003_observation_dedupe.sql"),
    ),
    (
        "0004_beliefs",
        include_str!("../migrations/0004_beliefs.sql"),
    ),
];

pub async fn run_migrations(pool: &PgPool) -> Result<(), sqlx::Error> {
    sqlx::query(
        "CREATE TABLE IF NOT EXISTS schema_migrations (version text PRIMARY KEY, applied_at timestamptz NOT NULL DEFAULT now())",
    )
    .execute(pool)
    .await?;

    let applied = sqlx::query_scalar::<_, String>("SELECT version FROM schema_migrations")
        .fetch_all(pool)
        .await?
        .into_iter()
        .collect::<HashSet<_>>();

    for &(version, migration_sql) in MIGRATIONS {
        if applied.contains(version) {
            continue;
        }

        let mut transaction = pool.begin().await?;
        sqlx::raw_sql(migration_sql)
            .execute(&mut *transaction)
            .await?;
        sqlx::query("INSERT INTO schema_migrations (version) VALUES ($1)")
            .bind(version)
            .execute(&mut *transaction)
            .await?;
        transaction.commit().await?;
    }

    Ok(())
}

pub const fn migration_count() -> usize {
    MIGRATIONS.len()
}
